File: dao-lesson/src/main/java/com/acme/recipes/data/entity/RecipeEntity.java
package com.acme.recipes.data.entity;

import io.realm.RealmList;
import io.realm.RealmModel;
import io.realm.annotations.PrimaryKey;
import io.realm.annotations.RealmClass;
import io.realm.annotations.Required;

@RealmClass
public class RecipeEntity implements RealmModel {

    @Required
    @PrimaryKey
    private String id;
    private String name;
    private String description;
    private int calories;
    private RealmList<IngredientEntity> ingredients;

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public int getCalories() {
        return calories;
    }

    public void setCalories(int calories) {
        this.calories = calories;
    }

    public RealmList<IngredientEntity> getIngredients() {
        return ingredients;
    }

    public void setIngredients(RealmList<IngredientEntity> ingredients) {
        this.ingredients = ingredients;
    }

    public RecipeEntity(String id, String name, String description, int calories) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.calories = calories;
    }

    // Empty constructor required by Realm.
    public RecipeEntity() {}

}

File: app/src/main/java/com/acme/recipes/database/entity/RecipeEntity.kt
package com.acme.recipes.database.entity

import com.acme.recipes.model.Ingredient
import com.acme.recipes.model.Recipe
import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey
import io.realm.annotations.Required

open class RecipeEntity() : RealmObject(), Recipe {

    @Required
    @PrimaryKey
    override var id: String = ""
    override var name: String = ""
    override var description: String = ""
    override var calories: Int = 0
    var ingredientEntities: RealmList<IngredientEntity> = RealmList()

    override val ingredients: List<Ingredient>?
        get() = ingredientEntities

    constructor(id: String, name: String, description: String, calories: Int) : this() {
        this.id = id
        this.name = name
        this.description = description
        this.calories = calories
    }

}